### PR TITLE
Add support for async

### LINKED ISSUE #20: async functions compatibility
It would be great if we could use log-derive with async functions. Adding `logfn` to an async function gives me the following:

```
error[E0728]: `await` is only allowed inside `async` functions and blocks
  --> src/models/video/mod.rs:94:9
   |
44 | #[logfn(ok = "INFO", err = "ERROR")]
   | ------------------------------------ this is not `async`
...
94 |         task::sleep(std::time::Duration::from_millis(500)).await;
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ only allowed inside `async` functions and blocks
```

Source Snippet:
```
#[logfn(ok = "INFO", err = "ERROR")]
pub async fn create_video_sdp() {
// ...
```

### PR DESCRIPTION
**Pull request by DmitrySamoylov**
The idea is to place function body inside
- async block when function is async
- immediately invoked closure when function is not async 

Some detals: https://users.rust-lang.org/t/nested-async-block-fn/44560/2

Closes #20

### CODE REVIEW

**Comment by elichai** [comment]
Tested and read on async blocks and this looks great :)
2 questions though:
1. Why does `async` needs to be behind a feature?
2. how do you feel about replacing `tokio` in the tests with `futures_executor::block_on`?

**Comment by DmitrySamoylov** [comment]
1. There is a check in Travis that we should build under older rust versions (`async` was introduced in 1.39). Did not touch that part but if we could drop support for older Rust versions then we don't need features.
2. Yeah that's fine. Any executor will do

**Comment by elichai** [comment]
>   1. There is a check in Travis that we should build under older rust versions (`async` was introduced in 1.39). Did not touch that part but if we could drop support for older Rust versions then we don't need features.

WIll this line fail to compile on older compilers? https://github.com/elichai/log-derive/pull/22/files#diff-914b24fcd759ea4189775083e2b4434cR9 if it doesn't we can replace it back to my `ExprClosure` and copy the asyncness of the function.

>   2. Yeah that's fine. Any executor will do

Feel free to do it(tokio is really big and I'd like to avoid it even for tests), otherwise I'll merge as is and change it later

**Comment by DmitrySamoylov** [comment]
1. Yep, this line will fail to compile. But also tests for async won't compile
2. Will do it in a moment

**Comment by elichai** [comment]
> 1. Yep, this line will fail to compile. But also tests for async won't compile

I'd rather feature gate tests and not actual logic if possible, most users don't really check features they just use the default. 
I'll look into maybe using `syn::ExprAsync` which I think should compile but will fail on the user side if he uses async functions with older compilers

**Comment by DmitrySamoylov** [comment]
So, the idea is to make an enum with `syn::ExprClosure` and `syn::ExprAsync` instead of raw `TokenStream`? This can work I think. Let me try :)

**Comment by DmitrySamoylov** [comment]
Think I finished force-pushing :smiley:

**Comment by elichai** [comment]
This is awesome!
I see you also decided to remove the call and added that to the expression itself :)
Too bad we need to specify so many default arguments (for a while I'm trying to decide between `Default::default()` and `token![SOMETHING]`) one small nit and i'll merge :)

File: tests/test_async.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#![cfg(feature = "async")]
```
Comment (elichai) [inline tests/test_async.rs]: this should be `async_test`

**Comment by elichai** [comment]
FYI, feel free to open an issue with some feedback about the UX and usability, it's pretty hard to know how people use this because on one hand it has 161 stars, but on the other 0 dependent crates.

**Comment by elichai** [comment]
Also @silverjam do you mind that the travis commit is now here? would you prefer a cherry-pick to include your contribution, or removing that commit from here and you'll rebase your PR afterwards? :)

File: tests/test_async.rs (hunk 1 above)
Comment (DmitrySamoylov) [reply tests/test_async.rs]: Good catch! Thanks!

**Comment by DmitrySamoylov** [comment]
Oh, I shouldn't have stolen @silverjam contribution
Cherry-picked those commits as a default option for @silverjam  :smiley:

**Comment by DmitrySamoylov** [comment]
> FYI, feel free to open an issue with some feedback about the UX and usability, it's pretty hard to know how people use this because on one hand it has 161 stars, but on the other 0 dependent crates.

Just started to use this crate and for now it looks very clean for me. All I could imagine is already here. Will add a feature request if I think of something :)

File: Cargo.toml [hunk 2]
```diff
@@ -14,12 +14,17 @@ include = [
     "Cargo.toml",
 ]
 
+[features]
+async = []
+async_test = ["async", "futures-executor"]
```
Comment (silverjam) [inline Cargo.toml]: Seems we can remove the `async` feature altogether now?

**Comment by silverjam** [comment]
> Also @silverjam do you mind that the travis commit is now here? would you prefer a cherry-pick to include your contribution, or removing that commit from here and you'll rebase your PR afterwards? :)

> Oh, I shouldn't have stolen @silverjam contribution
> Cherry-picked those commits as a default option for @silverjam

Thanks for keeping the attribution :-), but in this case it's small enough I don't mind either way.

File: Cargo.toml (hunk 2 above)
Comment (DmitrySamoylov) [reply Cargo.toml]: Oh, I removed it and then cherry-picked it back again.

**Comment by DmitrySamoylov** [comment]
@elichai Could you please bump version?

**Comment by elichai** [comment]
> @elichai Could you please bump version?

Done

**Pull request merged**